[B+C] Added BrewRecipe, PotionIngredientMeta, and BrewEndEvent. Added getPotions() to Brew(End)Event, which returns an array of Potion objects. Fixes Bukkit-2441 and BUKKIT-1604.

shinyquagsire23 opened the pull request:
#### The Issue:

Potions have always been the gray area of Bukkit. There's no way to make new potions, no way to add new brewing recipes, and there's no way to override the effects of a potion. 
#### Justification for this PR:

Many plugins attempt to override potion effects or add new effects by using the item's right click event. While a consume event has been added, this PR tackles the issue at it's core by allowing the overriding of a potion effect, so you could essentially give a potion a new effect or nullify it (ie, removing potion of invisibility by replacing it with water breathing). This PR also adds the BrewRecipe class, which acts as a way to increase the number of ingredients, and to modify their behavior through the use of the PotionIngredientMeta class, a wrapper around Minecraft's powerful potion engine. Lastly, this PR adds BrewEndEvent to enable the modification of the resulting potions. This allows for the removal/banning of potions right when they are added, as well as the modification of potions (ie, rename Potion of Invisibility to Potion of Water Breathing.) This PR also adds a getPotions method to Brew(End)Event which returns an array of Bukkit Potion objects for each potion
#### PR Breakdown:

BrewRecipe is tied into CraftBukkit much like the other recipes (Furnace, Crafting). A PotionIngredientMeta is instantiated defining precisely what the ingredient does. These effects include:
- Base Potion Effect, 0-15
- Extended
- Level II
- Splash
- Potion Base (Awkward, Thick, Mundane, or Water)
  TileEntityBrewing and SlotBrewing were modified to allow for the new ingredients added in the BrewRecipe to be inserted into the ingredient slot. TileEntityBrewing was then modified to take the new ingredient and retrieve the appropriate bitstring from the potion registry and would then start brewing. Once the brewing is done, BrewEndEvent is called to allow for final modifications of the potions.

Once the potion is complete and the player goes to drink it, PotionBrewer's modifications come in. These modifications allow for the new effects and durations to override the potion's default effects, or if the potion has no effects previously, to add new effects. This can also be done using BrewEndEvent by modifying the Potion's metadata, but one may be desireable over the other.
#### Plugin Example Code:

``` java
PotionIngredientMeta underwater = new PotionIngredientMeta(PotionType.ARTLESS, false, false, false, PotionBase.THICK); //13 is the base potion type, extended/levelI/gunpowder as boolean, and the potion's base

BrewRecipe recipe = new BrewRecipe(underwater, Material.INK_SACK); //Registers our ingredient with the PotionIngredientMeta
this.getServer().addRecipe(recipe); //Register the recipe like any other recipe

PotionRecipes.overrideEffect(8205, new PotionEffect(PotionEffectType.WATER_BREATHING, (60 * 3) * 20, 0)); //Override the effect of the resulting potion with the underwater breathing effect
```
#### Relevant PR(s)

The corresponding CraftBukkit pull request is Bukkit/CraftBukkit#943

Unh0lyTigg commented on src/main/java/org/bukkit/event/inventory/BrewEndEvent.java:
```diff
@@ -0,0 +1,50 @@
+    private boolean cancelled;
+    private ItemStack ingredient;
+
+    public BrewEndEvent(Block brewer, BrewerInventory contents, ItemStack ingredient) {
```
~~The ItemStack "ingredient" is declared as a parameter in this constructor, but it is never used within the constructor to set the class-wide level ItemStack "ingredient"~~

jrtc27 commented:
Don't use strings (nor just numbers) for the bitField.

jrtc27 commented on src/main/java/org/bukkit/event/inventory/BrewEvent.java:
```diff
@@ -20,6 +21,11 @@ public BrewerInventory getContents() {
     }
 
+    public Potion[] getPotions()
+    {
```
Line above

jrtc27 replied:
Still needs fixing.

jrtc27 replied:
Formatting still wrong...

jrtc27 commented on src/main/java/org/bukkit/event/inventory/BrewEvent.java:
```diff
@@ -20,6 +21,11 @@ public BrewerInventory getContents() {
 
+    public Potion[] getPotions()
+    {
+    	return new Potion[] { Potion.fromItemStack(contents.getContents()[0]), Potion.fromItemStack(contents.getContents()[1]), Potion.fromItemStack(contents.getContents()[2]) };
```
Indent with 4 spaces

shinyquagsire23 commented:
Fixed the two errors you both reported. Thank you for spotting those. :)

@jrtc27 
Actually, NMS uses Strings for the bit modification, so using a string here was best to avoid too much (needless) modification to the NMS coding. Not to mention the Strings have checking operators & and !, which the attached JavaDoc explains.

LB-- commented:
Don't forget to squash the commits ;)

shinyquagsire23 commented:
Shoot, screwed up my Git repo while renaming some commits. -.-

EDIT:
Fixed, and squished those pesky commits! :D

shinyquagsire23 committed:
Added BrewRecipe, PotionRecipes, and BrewEndEvent. Fixes Bukkit-2441 and BUKKIT-1604. Instead of using an Input+Potion=NewPotion System, which isn't quite robust enough to immitate the system used for normal potions, it works by adding a class wrapper to recieve arguments about which base potion to use (Awkward, Thick, or Mundane), which additions (gunpowder, lv II, extended), which potion type it will create (accepts PotionType or int), as well as the ingredient that will add these different arguments. After a recipe is added, you can call PotionRecipes.overrideEffect(...) to override a potion effect. Added PotionRecipes as a class to store the different overrides and recipes. For more information as to how this works, you can check the JavaDoc attached to the BrewRecipe and PotionBitSet files.

Wolvereness commented:
Just because NMS uses strings with complex notation, doesn't mean Bukkit should. The fact that the notation is complex difficult to document is reason to not put it in Bukkit, instead wrapping an API around it.

shinyquagsire23 commented:
Ah, I'll work something out like that. I'll probably make a class called PotionBitSet that accepts a PotionEffect(Type), but also allows the manual bit setting/checking. Then I'll fix up something to spit out the bit notation.

shinyquagsire23 commented:
Squishy squashed down the bugfix commits. :smiley:

jrtc27 commented:
The formatting in both pull requests is still dodgy - go to the "Fles Changed" tab and just go through line by line checking the formatting. Also your squashing doesn't seem to have worked...

jrtc27 commented on src/main/java/org/bukkit/event/inventory/BrewEndEvent.java:
```diff
@@ -0,0 +1,58 @@
+        if(contents.getContents()[1] != null)
+            potions[0] = Potion.fromItemStack(contents.getContents()[1]);
+        if(contents.getContents()[2] != null)
+            potions[0] = Potion.fromItemStack(contents.getContents()[2]);
```
Do you really mean `potions[0]`?

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,592 @@
+    public int baseID = 0;
+
+    public enum PotionBase {
+        WATER, AWKWARD, THICK, MUNDANE,
```
Not how Bukkit formats enums - look at ChatColor

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,592 @@
+    }
+
+    public enum PotionBit {
+        ZERO(0), ONE(1), TWO(2), THREE(3), EFFECT(new int[] { 0, 1, 2, 3 }), LEVELII(5), EXTENDED(6), DRINKABLE(13), THROWABLE(14), AWKWARD(4), THICK(5), MUNDANE(6), UNKNOWN(new int[] { 7, 8, 9, 10, 11, 12, 15 });
```
As above

shinyquagsire23 commented:
Sorry I didn't fix the enum's earlier. I was busy with school projects and Thanksgiving, but now it's fixed. I also attached some JavaDocs as well.

riking commented:
Can you go ahead and do a `git rebase -i bukkit/master` and squash these all into 1 commit?

Edit: :)

shinyquagsire23 commented:
Aaaaaand... Squashed.

jrtc27 commented on src/main/java/org/bukkit/inventory/BrewRecipe.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Returns the bit modifier string passed to NMS
+     * 
+     * @return
```
Incomplete JavaDoc EDIT: not any more!

jrtc27 commented on src/main/java/org/bukkit/inventory/BrewRecipe.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Returns the bit modifier as a PotionBitSet
+     * 
+     * @return
```
Also incomplete EDIT: Not any more!

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,671 @@
+            return bits.clone();
+        }
+
+        public int getBit() {
```
Is this even necessary?

shinyquagsire23 replied:
Yes, if you want to use the bit setters/getters in the class you can use PotionBit.AWKWARD.getBit()
Plus, I usually add getters if there's a setter. Good practice in case myself or others need it at one point or another.

riking replied:
Maybe getBit() needs a javadoc - "Returns -1 on arrays. This is generally only to be used on the constants."

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,671 @@
+            if (this != UNKNOWN)
+                return getBits()[0]; // Returns 0 on EFFECT
+            else
+                return 0;
```
Maybe -1 instead to differentiate it from EFFECT/ZERO?

jrtc27 commented:
Some formatting is still wrong - curly braces should **never** be on a new line - and a couple of your JavaDoc entries are missing documentation for what the function return.

shinyquagsire23 commented:
Fixed the JavaDoc and the curly brace. Sorry about that, I'm just so used to newline curly braces. I guess I forgot to reformat that section with Eclipse.

I'll squish the commits later. I'm kinda coding on the fly at the moment (using GitHub's edit function), and I'm not at a computer I can squish those from.

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,671 @@
+    private boolean[] bitarray = new boolean[16];
+    private boolean[] setfalse = new boolean[16];
+    private boolean[] bitcheckif = new boolean[16];
+    private boolean[] bitchecknot = new boolean[16];
```
Please please please use camel case

shinyquagsire23 commented:
Fixed and Re-Squashed :)

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,671 @@
+            bits = "0" + bits;
+        int count = 0;
+        for (int i = 3; i >= 0; i--)
+        {
```
Wrong line :P

jrtc27 commented on src/main/java/org/bukkit/potion/PotionBitSet.java:
```diff
@@ -0,0 +1,671 @@
+            bitArray[13] = true;
+
+        switch (base) {
+        case AWKWARD:
```
Indentation - `case` should be 4 more than switch, and everything inside it a further 4 more.

shinyquagsire23 commented:
Fixed again. Squashed again

shinyquagsire23 commented:
I decided that it would be best if I moved PotionRecipes.java to org.bukkit.potion from org.bukkit.craftbukkit.potion to avoid having to import a class from CraftBukkit, and because doing so it a pain in the rear anyways. I'm not even sure why I put it in CraftBukkit anyways. :p

riking commented:
PotionRecipes needs to be javadoc'd and removed from https://github.com/Bukkit/CraftBukkit/issues/943.

Additionally, as I suggested over there, you should use `Custom` instead of `New` in PotionRecipes - it more accurately conveys their effects.

Also rebase after javadocing

THDigi commented:
Hmm, how does this deal with the fact that clients can't place all types of items in the brewer's ingredient slot ? They can only place the original ingredients... gunpowder, redestone, etc, but not gold ingots for example.

shinyquagsire23 commented:
Fixed the said issue, and added a better description of the PR's goals.

THDigi: That was actually a tough part during my development of this. I discovered that it worked on the server side when I allowed new ingredients from SlotBrewing.java (in NMS). Since the server side part works, well, server side, I had to force an update via Player.updateInventory method of each of the viewers of the GUI, thus making a seamless update to the client to allowing restricted items.

shinyquagsire23 committed:
Renamed PotionBitSet to PotionIngredientMeta to avoid using NMS terms and prevent confusion. Added the unused potions to PotionType to allow their usage in PotionIngredientMeta. These unused potions may be replaced by their new counterparts when a new potion is released.

shinyquagsire23 committed:
Fixed some Javadocs and finished renaming.

riking commented on src/main/java/org/bukkit/event/inventory/BrewEvent.java:
```diff
@@ -36,4 +48,4 @@ public HandlerList getHandlers() {
         return handlers;
     }
-}
+}
```
You removed the newline at end of file.

note to self: strikethrough once fixed, delete others

riking commented on src/main/java/org/bukkit/event/inventory/BrewEvent.java:
```diff
@@ -20,6 +21,17 @@ public BrewerInventory getContents() {
         return contents;
     }
 
+    public Potion[] getPotions() {
```
This should be an immutable list of ItemStack instead

riking commented on src/main/java/org/bukkit/inventory/BrewRecipe.java:
```diff
@@ -0,0 +1,91 @@
+        return bitSet;
+    }
+
+}
```
Needs newline at end of file

note to self: delete once fixed (top one will be kept)

riking commented on src/main/java/org/bukkit/potion/PotionRecipes.java:
```diff
@@ -0,0 +1,162 @@
+    public static HashMap<Integer, String> getRecipes() {
+        return recipes;
+    }
+}
```
Needs newline at end of file

note to self: delete once fixed (top one will be kept)

riking commented:
We need to change this to use the new ItemMeta system that @Wolvereness introduced.

Add more methods to https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/inventory/meta/PotionMeta.java

~~In the same vein as getScoreboardManager() and getItemFactory(), add Bukkit.getPotionBrewer() / Server.getPotionBrewer().~~

~~The contract for PotionBrewer should be such that a plugin can get an awkward potion by calling `getServer().getPotionBrewer().getBrewResult(Material.NETHER_WART, new MaterialData(Material.POTION, 0));` (assuming I wrote that right).~~

~~A plugin should be able to do PotionBrewer.addRecipe(), then subsequent calls to getBrewResult() would follow that recipe.~~

~~The BrewingStand should use the PotionBrewer when deciding the result.~~

See my [other comment](https://github.com/Bukkit/CraftBukkit/pull/943#issuecomment-16467909)
